### PR TITLE
REP-2185 Implement Self Validation Functionality

### PR DESCRIPTION
**Pull request by dmnjohns**

### CODE REVIEW

File: repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala [hunk 1]
```diff
@@ -81,8 +81,8 @@ class KeystoneRequestHandler(identityServiceUri: String, akkaServiceClient: Akka
               case Failure(f) =>
                 Failure(IdentityCommunicationException("Token not found in identity response during admin authentication", f))
             }
-          case SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS =>
-            Failure(OverLimitException(buildRetryValue(serviceClientResponse), "Rate limited when getting admin token"))
+          case statusCode@(SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS) =>
```
Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala]: I was unaware of this syntax for binding a matching pattern to a variable. Cool. :+1:

File: repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala [hunk 2]
```diff
@@ -145,7 +147,11 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
             case Failure(e: UnauthorizedEndpointException) => Reject(SC_FORBIDDEN, Some(e.getMessage))
             case Failure(e: OverLimitException) =>
               response.addHeader(HttpHeaders.RETRY_AFTER, e.retryAfter)
-              Reject(HttpServletResponse.SC_SERVICE_UNAVAILABLE, Some(e.getMessage))
+              if (isSelfValidating) {
+                Reject(e.statusCode, Some(e.getMessage))
+              } else {
+                Reject(HttpServletResponse.SC_SERVICE_UNAVAILABLE, Some(e.getMessage))
```
Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: All of the `HttpServletResponse._` are already imported. So there is no need to explicitly reference them. This isn't exactly part of this change, but they should probably be fixed.

File: repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala [hunk 3]
```diff
@@ -145,7 +147,11 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
...
+              }
             case Failure(e) if e.getCause.isInstanceOf[AkkaServiceClientException] && e.getCause.getCause.isInstanceOf[TimeoutException] =>
               Reject(HttpServletResponse.SC_GATEWAY_TIMEOUT, Some(s"Call timed out: ${e.getMessage}"))
```
Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: (2/4)

File: repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala [hunk 4]
```diff
@@ -314,18 +330,18 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
...
-    def getEndpoints(getAdminToken: Boolean => Try[String], authToken: String, validToken: ValidToken): Try[EndpointsData] = {
+    def getEndpoints(authToken: String, validToken: ValidToken): Try[EndpointsData] = {
       logger.trace(s"Getting endpoints for: $authToken")
 
       // todo: extract the "make call with admin token and retry" logic since that pattern is used elsewhere
       Option(datastore.get(s"$ENDPOINTS_KEY_PREFIX$authToken").asInstanceOf[EndpointsData]) match {
         case Some(endpointsData) => Success(endpointsData)
         case None =>
-          getAdminToken(false) flatMap { adminToken =>
+          getValidatingToken(authToken, force = false) flatMap { adminToken =>
```
Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: Why use one named parameter and one not, but still in the method definition order?

Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: Was this due to the IntelliJ IDEA static analysis highlight?

Comment (dmnjohns) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: It was. Naming both seems unnecessary, although so does naming only the boolean. Still, that's what IntelliJ likes, and I like not having IntelliJ whine at me. I don't mind changing this, nor do I mind leaving it as is, just let me know what you prefer.

Comment (wdschei) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: I am in the make IntelliJ happy camp. They have a few things they flag that are a little over the top, but I don't think they are so absurd as to put the effort into policing configs for the entire team. I call it good. ;-)

Comment (dmnjohns) [inline repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala]: I think so, too. I'll leave it as-is, then.

**Pull request merged**